segmentstore/storage/src/test/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorageTests.java

```diff
@@ -1371,6 +1371,102 @@ public void testBaseConcatWithDefragWithMinMaxLimits() throws Exception {
+        testConcatUsingAppendsAfterWriteFailure(maxRollingSize,
+                new long[] {5},
+                new long[] {1, 2, 3, 4},
+                new int[]  {0, 1, 2, 3},  // add garbage to these chunks
+                new long[] {5, 1, 2, 3, 4},
+                15);
+    }
+
+    private void testConcatUsingAppendsAfterWriteFailure(long maxRollingSize,
+                                                         long[] targetLayoutBefore,
+                                                         long[] sourceLayout,
+                                                         int[] chunksWithGarbageIndex,
+                                                         long[] targetLayoutAfter,
+                                                         long expectedLength) throws Exception {
+        String targetSegmentName = "target";
+        String sourceSegmentName = "source";
+        SegmentRollingPolicy policy = new SegmentRollingPolicy(maxRollingSize); // Force rollover after every 20 byte.
+        ChunkedSegmentStorageConfig config = ChunkedSegmentStorageConfig.DEFAULT_CONFIG.toBuilder()
+                .maxSizeLimitForConcat(100)
+                .minSizeLimitForConcat(100)
+                .build();
+
+        @Cleanup
+        TestContext testContext = getTestContext(config);
+        ((AbstractInMemoryChunkStorage) testContext.chunkStorage).setShouldSupportConcat(true);
+
+        // Create target
+        testContext.insertMetadata(targetSegmentName, maxRollingSize, 1, targetLayoutBefore);
+
+        // Create source
+        testContext.insertMetadata(sourceSegmentName, maxRollingSize, 1, sourceLayout);
+        val hSource = testContext.chunkedSegmentStorage.openWrite(sourceSegmentName).get();
+        testContext.chunkedSegmentStorage.seal(hSource, null).get();
+
+        // Add some garbage data at the end of last chunk
+        val lastChunkMetadata = TestUtils.getChunkMetadata(testContext.metadataStore,
+                TestUtils.getSegmentMetadata(testContext.metadataStore, targetSegmentName).getLastChunk());
+        testContext.chunkStorage.write(ChunkHandle.writeHandle(lastChunkMetadata.getName()), lastChunkMetadata.getLength(), 1, new ByteArrayInputStream(new byte[1]));
+
+        // Write some garbage at the end.
+        val sourceList = TestUtils.getChunkList(testContext.metadataStore, sourceSegmentName);
+        for (int i : chunksWithGarbageIndex) {
+            // Append some data to the last chunk to simulate partial write during failure
+            val chunkMetadata = TestUtils.getChunkMetadata(testContext.metadataStore, sourceList.get(i).getName());
+            testContext.chunkStorage.write(ChunkHandle.writeHandle(chunkMetadata.getName()), chunkMetadata.getLength(), 1, new ByteArrayInputStream(new byte[1]));
+        }
+        val hTarget = testContext.chunkedSegmentStorage.openWrite(targetSegmentName).get();
+        testContext.chunkedSegmentStorage.concat(hTarget, 5, sourceSegmentName, null).join();
```

Comment:
Is this validating the `InvalidOffsetException`, right? As it is not validating that such exception is thrown, it means that the  defragmentation will continue working as usual, skipping any number of such failures.

Reply:
Yes. Here we are writing additional garbage data to chunk directly via `ChunkStorage` without SLTS knowing about it. And it tries to write at the wrong offset.
In this case the `ChunkStorage` contract requires it to throw `InvalidOffsetException` when write operation is called by SLTS with wrong offset - which it throws.

This way the test recreates the exact situation we are trying to fix.